### PR TITLE
disable deepcopy for list action

### PR DESCRIPTION
**Pull request by CharlesQQ**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

When the node CPU usage rate is almost over 90%, it seems that deepcopy cost the most time:

```
I1113 14:36:11.161745      18 overridemanager.go:181] applyNamespacedOverrides, list configmap-beta-moa-demo-prod-681,  cluster member-cluster1, cost: 0.980144234
I1113 14:36:12.445315      18 overridemanager.go:181] applyNamespacedOverrides, list configmap-beta-moa-demo-prod-681,  cluster member-cluster2, cost: 0.811084427
```
image

After turning off list deepcopy, the list time is reduced to 0.1s or even lower.
```
I1113 15:43:39.212238      17 overridemanager.go:182] applyNamespacedOverrides, list beta-moa-demo-prod-283,  cluster member-cluster1, cost: 0.040210384
I1113 15:43:39.251189      17 overridemanager.go:182] applyNamespacedOverrides, list beta-moa-demo-prod-97,  cluster member-cluster2, cost: 0.018000362
I1113 15:43:39.314731      17 overridemanager.go:182] applyNamespacedOverrides, list beta-moa-demo-prod-283,  cluster member-cluster2, cost: 0.023638913
```
image

**Which issue(s) this PR fixes**:
part of  https://github.com/karmada-io/karmada/issues/5790

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
`Performance`: `karmada-controller-manager`: Significant performance improvements have been achieved by reducing deepcopy operations during list processes:
- ​​Binding Controller​​: Response time reduced by 50%
​​- Dependencies Controller​​: Execution time improved 5× faster 
​​- Detector Controller​​: Processing time cut by 50% 
```

### CODE REVIEW

**Comment by zach593** [comment]
Regarding this performance optimization, which clearly introduces side effects, I have two suggestions:

1. How much of a performance improvement will it bring? Could you add benchmarks for these calls along with e2e performance comparison?

2. Is it possible to add some mechanisms to mitigate these side effects? For example, adding unit tests or e2e tests so that if someone later introduces code that triggers this side effect (i.e., directly modifying the informer cache), our test suite can detect such behavior.

**Comment by CharlesQQ** [comment]
> How much of a performance improvement will it bring? Could you add benchmarks for these calls along with e2e performance comparison?

Through golang pprof analysis, it is found that during the restart period, deep copy will have obvious CPU time loss. After removing it, this part of the loss disappeared.
Maybe this change is difficult to compare through e2e testing. Do you have any good testing suggestions?

**Comment by CharlesQQ** [comment]
if seems that deepcopy cost the most time:
image

image
image

image

After turning off list deepcopy, the list time is reduced to 0.1s or even lower.
```
I1113 15:43:39.212238      17 overridemanager.go:182] applyNamespacedOverrides, list beta-moa-demo-prod-283,  cluster member-cluster1, cost: 0.040210384
I1113 15:43:39.251189      17 overridemanager.go:182] applyNamespacedOverrides, list beta-moa-demo-prod-97,  cluster member-cluster2, cost: 0.018000362
I1113 15:43:39.314731      17 overridemanager.go:182] applyNamespacedOverrides, list beta-moa-demo-prod-283,  cluster member-cluster2, cost: 0.023638913
```
image

**Comment by CharlesQQ** [comment]
pls take a look at the e2e performance comparison  @zach593

**Comment by zach593** [comment]
> 2\. Is it possible to add some mechanisms to mitigate these side effects? For example, adding unit tests or e2e tests so that if someone later introduces code that triggers this side effect (i.e., directly modifying the informer cache), our test suite can detect such behavior.

There is an env `KUBE_CACHE_MUTATION_DETECTOR`, which may cause panic when the informer cache is illegally modified, I think if we can enable this environment when running e2e tests, I guess it could eliminate the side effects of disable deepcopy.

**Comment by CharlesQQ** [comment]
> Reducing the time taken for an operation by an order of magnitude is great, but I think it would be more helpful to see how much of an overall improvement the change made, especially for an optimization that has potential side effects.


I will test in a real test environment and give optimization data

**Comment by CharlesQQ** [comment]
> here is an env KUBE_CACHE_MUTATION_DETECTOR, which may cause panic when the informer cache is illegally modified, I think if we can enable this environment when running e2e tests, I guess it could eliminate the side effects of disable deepcopy.

KUBE_CACHE_MUTATION_DETECTOR=true take effect in go-client sharedinformer, does it also work in controller-runtime? If so, this environment variable should be added in e2e tests

**Comment by zach593** [comment]
> `KUBE_CACHE_MUTATION_DETECTOR=true` take effect in go-client sharedinformer, does it also work in controller-runtime? 

https://github.com/karmada-io/karmada/blob/0713e4b9be36e4b079f1dc30ac6570d17c9e6451/vendor/sigs.k8s.io/controller-runtime/pkg/cache/internal/informers.go#L59-L61


The answer is yes, it also applies to controller-runtime, because controller-runtime also uses client-go informer.

**Comment by CharlesQQ** [comment]
> A new concern just popped into my mind: does disabling deepCopy consistently lead to data being in a state of race conditions?

I don't understand what this sentence means, can you give me an example?

**Comment by CharlesQQ** [comment]
4000 workload, 8000 rb, 12C cpu:

before optimization, it tooks 9 minutes; after optimization, it takes 4 minutes and 30 seconds

image

image

**Comment by zach593** [comment]
I'm wondering if there will be a data race between the reading thread and the informer writing thread. This is something I suddenly realized, and I didn't think about it very seriously. Since I raised this question, I did some research and the result is **it should not cause a data race**.

https://github.com/karmada-io/karmada/blob/9ea0382293c80fdf8cf13da914136dfc4df3ec51/vendor/k8s.io/client-go/tools/cache/controller.go#L548-L557

From the code above, you can see that when writing to the cache, the entire variable is replaced with the new object from the `deltas` .If the read thread holds an object, when the write thread updates the cache, the object held by the read thread and the cache will no longer have any connection.

In other words, if the reading thread modifies the held object in the `*a.xxx = yyy` way, then this will cause a data race, and will also pollute the cache and objects already held by other reading threads. The object read from the cache later will also be polluted. Pollution will continue until the informer updates the object.

**Comment by CharlesQQ** [comment]
> In other words, if the reading thread modifies the held object in the *a.xxx = yyy way, then this will cause a data race, and will also pollute the cache and objects already held by other reading threads. The object read from the cache later will also be polluted. Pollution will continue until the informer updates the object.

list action disabling deepCopy, must make sure the list item will not be modified, or make a deepcopy of it before modifying it.

**Comment by CharlesQQ** [comment]
detector controller:   takes 2 minutes and 30 seconds
image

after disable deepcopy for list action, it takes 1 minute and 15 seconds
image

**Review by RainbowMango — COMMENTED** [review]
Disable deepcopy indeed can improve performance, but should be used very cautiously.

/assign

File: pkg/controllers/binding/binding_controller.go [hunk 1]
```diff
@@ -190,7 +191,7 @@ func (c *ResourceBindingController) newOverridePolicyFunc() handler.MapFunc {
 		}
 
 		bindingList := &workv1alpha2.ResourceBindingList{}
-		if err := c.Client.List(ctx, bindingList); err != nil {
+		if err := c.Client.List(ctx, bindingList, &client.ListOptions{UnsafeDisableDeepCopy: ptr.To(true)}); err != nil {
```
Comment (RainbowMango) [inline pkg/controllers/binding/binding_controller.go:194]: Does it make sense to add a `namespace` filter? Like:
```diff
                bindingList := &workv1alpha2.ResourceBindingList{}
-               if err := c.Client.List(ctx, bindingList); err != nil {
+               if err := c.Client.List(ctx, bindingList, &client.ListOptions{UnsafeDisableDeepCopy: ptr.To(true), Namespace: namespace}); err != nil {
```

Comment (RainbowMango) [reply pkg/controllers/binding/binding_controller.go:194]: `bindingList`  --> `readonlyBindingList`?

File: pkg/dependenciesdistributor/dependencies_distributor.go [hunk 2]
```diff
@@ -175,8 +176,10 @@ func (d *DependenciesDistributor) reconcileResourceTemplate(key util.QueueKey) e
 	klog.V(4).Infof("DependenciesDistributor start to reconcile object: %s", resourceTemplateKey)
 	bindingList := &workv1alpha2.ResourceBindingList{}
```
Comment (RainbowMango) [inline pkg/dependenciesdistributor/dependencies_distributor.go:177]:
```suggestion
	readonlyBindingList := &workv1alpha2.ResourceBindingList{}
```

**Review by RainbowMango — COMMENTED** [review]
Generally looks good to me. 
But I think we should rename variable name or add more comments to notice the object can not be mutated.

File: pkg/controllers/binding/binding_controller.go (hunk 1 above)
Comment (CharlesQQ) [reply pkg/controllers/binding/binding_controller.go:194]: > Does it make sense to add a namespace filter

Yes

**Comment by CharlesQQ** [comment]
>  But I think we should rename variable name or add more comments to notice the object can not be mutated.

Controller-runtime has a detailed description of `UnsafeDisableDeepCopy`, which is very clear. Is there no need to add it?

https://github.com/karmada-io/karmada/blob/65e7f74c991631a5888f5176342af118f4d5712e/vendor/sigs.k8s.io/controller-runtime/pkg/client/options.go#L495-L499

**Review by RainbowMango — APPROVED** [review]
/lgtm
/approve
